include/dbgroup/lock/optiql.hpp

```diff
@@ -0,0 +1,298 @@
+     */
+    constexpr OptGuard(  //
+        OptiQL *dest,
+        const uint32_t ver)
+        : dest_{dest}, ver_{ver}
+    {
+    }
+
+    constexpr OptGuard(const OptGuard &) = default;
+    constexpr OptGuard(OptGuard &&) noexcept = default;
+
+    constexpr auto operator=(const OptGuard &) noexcept -> OptGuard & = default;
+    constexpr auto operator=(OptGuard &&) noexcept -> OptGuard & = default;
+
+    /*##########################################################################
+     * Public destructors
+     *########################################################################*/
+
+    ~OptGuard() = default;
+
+    /*##########################################################################
+     * Public getters
+     *########################################################################*/
+
+    /**
+     * @return false.
+     */
+    constexpr explicit
+    operator bool() const
+    {
+      return false;
+    }
+
+    /**
+     * @return The version when this guard was created.
+     */
+    [[nodiscard]] constexpr auto
+    GetVersion() const  //
+        -> uint32_t
+    {
+      return ver_;
+    }
+
+    /*##########################################################################
+     * Public APIs
+     *########################################################################*/
+
+    /**
+     * @retval true if a target version does not change from an expected one.
+     * @retval false otherwise.
+     */
+    [[nodiscard]] auto VerifyVersion()  //
+        -> bool;
+
+   private:
+    /*##########################################################################
+     * Internal member variables
+     *########################################################################*/
+
+    /// @brief The address of a target lock.
+    OptiQL *dest_{};
+
+    /// @brief A version when creating this guard.
+    uint32_t ver_{};
+  };
+
+  /*############################################################################
+   * Public constructors and assignment operators
+   *##########################################################################*/
+
+  constexpr OptiQL() = default;
+
+  OptiQL(const OptiQL &) = delete;
+  OptiQL(OptiQL &&) = delete;
+
+  auto operator=(const OptiQL &) -> OptiQL & = delete;
+  auto operator=(OptiQL &&) -> OptiQL & = delete;
+
+  /*############################################################################
+   * Public destructors
+   *##########################################################################*/
+
+  ~OptiQL() = default;
+
+  /*############################################################################
+   * Public APIs
+   *##########################################################################*/
+
+  /**
+   * @return An empty guard instance with the current version value.
+   *
+   * @note This function does not give up reading a version value and continues
+   * with spinlock and back-off.
+   */
+  [[nodiscard]] auto GetVersion()  //
```

Comment:
見逃してましたが，メンバ変数の変更なしなので `const` ですね．cppファイルの方も同様の変更がいるはずです．

```suggestion
  [[nodiscard]] auto GetVersion() const  //
```